pkg/threescale/backend/manager.go

```diff
@@ -0,0 +1,367 @@
+	if transactions == nil || len(transactions) == 0 {
+		return api.Transaction{}, fmt.Errorf("transaction must be non-nil and non-empty")
+	}
+
+	transactionCopy := make([]api.Transaction, len(transactions))
+	copy(transactionCopy, transactions)
+	transactionCopy[0].Metrics = nil
+
+	return transactionCopy[0], nil
+}
+
+func (b *Backend) applicationFromResponse(resp threescale.AuthorizeResult) Application {
+	limitCounter := make(map[string]map[api.Period]*Limit)
+
+	for metric, report := range resp.GetUsageReports() {
+		l := &Limit{
+			current: report.CurrentValue,
+			max:     report.MaxValue,
+		}
+		limitCounter[metric] = make(map[api.Period]*Limit)
+		limitCounter[metric][report.PeriodWindow.Period] = l
+	}
+
+	return Application{
+		LastResponse:  resp,
+		LimitCounter:  limitCounter,
+		UnlimitedHits: make(map[string]int),
+	}
+}
+
+func (b *Backend) computeAffectedMetrics(hierarchy api.Hierarchy, metrics api.Metrics) api.Metrics {
+	for parent, children := range hierarchy {
+		for k, v := range metrics {
+			if contains(k, children) {
+				if _, known := metrics[parent]; known {
+					metrics[parent] += v
+				} else {
+					metrics.Add(parent, v)
+				}
+			}
+		}
+	}
+	return metrics
+}
+
+// isAuthorized works on a copy of the passed Application and will inform the caller if limits have
+// been breached in the provided affected metrics
+func (b *Backend) isAuthorized(application Application, affectedMetrics api.Metrics) (Application, bool) {
+	copiedApplication := application.deepCopy()
+	commitChanges := true
+
+out:
+	for metric, incrementBy := range affectedMetrics {
+		cachedValue, ok := copiedApplication.LimitCounter[metric]
+		if ok {
+			for _, limit := range cachedValue {
+				newValue := limit.current + incrementBy
+				if newValue > limit.max {
+					commitChanges = false
+					break out
+				}
+				limit.current = newValue
+			}
+		} else {
+			// has no limits so just cache the value for reporting purposes
+			current, ok := copiedApplication.UnlimitedHits[metric]
+			if ok {
+				copiedApplication.UnlimitedHits[metric] = current + incrementBy
+				break
+			}
+			copiedApplication.UnlimitedHits[metric] = incrementBy
+		}
+	}
+	return copiedApplication, commitChanges
+}
+
+func (b *Backend) localReport(cacheKey string, application Application, affectedMetrics api.Metrics) Application {
+	copiedApplication := application.deepCopy()
+
+	for metric, incrementBy := range affectedMetrics {
+		cachedValue, ok := copiedApplication.LimitCounter[metric]
+		if ok {
+			for _, limit := range cachedValue {
+				newValue := limit.current + incrementBy
+				limit.current = newValue
+			}
+		} else {
+			// has no limits so just cache the value for reporting purposes
+			current, ok := copiedApplication.UnlimitedHits[metric]
+			if ok {
+				copiedApplication.UnlimitedHits[metric] = current + incrementBy
+				break
+			}
+			copiedApplication.UnlimitedHits[metric] = incrementBy
```

Comment:
This is pretty much the code above, and we can probably extract it into its own function.

Reply:
yup